### PR TITLE
fixes #203, block assignment tests and doc

### LINKED ISSUE #203: figure out what's up with assign(Eigen<Block>) without reference
assign.hpp looks wrong.  Check it out.

### PR DESCRIPTION
**Pull request by bob-carpenter**
#### Submission Checklist

- [x] Run unit tests: `./runTests.py test/unit`
- [x] Run cpplint: `make cpplint`
- [x] Declare copyright holder and open-source license: see below

#### Summary:

The code was correct as it was originally, but I added some doc to make that clear and an additional straight-up block test.

#### Intended Effect:

Make sure nobody gets confused about this in the future.

#### How to Verify:

Unit tests.

#### Side Effects:

No---only adds test and doc.

#### Documentation:

n/a

#### Reviewer Suggestions: 

anyone

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company):

Columbia University

By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses:
- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
- Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

### CODE REVIEW

**Comment by bob-carpenter** [comment]
This one should be super easy to review.  I just added extra doc and tests to make it clear why the odd programming pattern of not calling the block by constant reference is there.

**Comment by wds15** [comment]
First time reviewing; still fighting with github. Anyway: The added comment and test makes sense. I am not quite in favor of the renaming of RHS to T template name, but fine with me if this is a newer policy I am not aware of.

**Comment by bob-carpenter** [comment]
OK---doesn't make sense to have them be inconsistent.  I don't like these long all caps name---for one, they violate the naming convention for types.  So if I do change them and everyone insists on rhs and lhs, choose between `T_rhs`, `Trhs` or `Rhs` --- we shouldn't be using all caps because those are reserved for ordinary constants.

**Comment by bob-carpenter** [comment]
@wds15 --- what you need to do is go into discussion and start commenting inline.  It'll ask if you want to start a review, then you can wrap all the comments up together.  Then merging will be blocked until you approve fixing changes.

**Comment by bob-carpenter** [comment]
We don't want to merge until at least one independent reviewer has put in an actual review.

**Comment by wds15** [comment]
Then what about

`T_rhs` and `T_lhs` ?

This would be consistent with the ODE code which I am familiar with (where we have `T_param`).

Do we want to make the file consistent in this regard and change all the other occurrences?

**Comment by bob-carpenter** [comment]
I can update it and repush.  I'm dealing with emergency Eigen version and C++ compiler issues right now, but will get to it soon.

File: stan/math/prim/mat/fun/assign.hpp [hunk 1]
```diff
@@ -120,24 +120,29 @@ namespace stan {
...
      * column shape template parameters.
      *
+     * <p>The left-hand side is intentionally not a reference, because
+     * that won't match generally enough;  instead, a non-reference is
+     * used, which still holds onto a reference to the contained
+     * matrix and thus still updates the appropriate values.
+     *
      * @tparam LHS Type of matrix block elements.
-     * @tparam RHS Type of right-hand side matrix elements.
+     * @tparam T Type of right-hand side matrix elements.
```
Comment (syclik) [inline stan/math/prim/mat/fun/assign.hpp]: @wds15 suggests these be `T_lhs` and `T_rhs`

File: stan/math/prim/mat/fun/assign.hpp [hunk 2]
```diff
@@ -120,24 +120,29 @@ namespace stan {
...
      * @tparam R Row shape for right-hand side matrix.
      * @tparam C Column shape for right-hand side matrix.
      * @param x Left-hand side block view of matrix.
      * @param y Right-hand side matrix.
      * @throw std::invalid_argument if sizes do not match.
      */
-    template <typename LHS, typename RHS, int R, int C>
+    template <typename LHS, typename T, int R, int C>
```
Comment (syclik) [inline stan/math/prim/mat/fun/assign.hpp]: `T_lhs` and `T_rhs`

**Pull request merged**